File: bridges/nvda/addon/globalPlugins/nvdaMcpBridge/domain/controllers/commands/command_handler.py
# nvdaMcpBridge domain -- CommandHandler: the per-command controller interface.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
# ROLE: the interface every command handler implements.
# USED BY: the Session and registry.py.
# A handler fails a command by raising; the Session turns that into an error Response.

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
	from .... import protocol
	from .session_context import SessionContext


#: The longest any blocking command may hold the session thread; it must stay inside the 120 s
#: inactivity window, which a blocking handler does not refresh.
MAX_POLL_TIMEOUT: float = 110.0


class CommandError(Exception):
	"""A handler-level failure that becomes an error Response."""


class CommandHandler(ABC):
	#: ``ping`` sets this False: it proves liveness, not that the agent is still testing.
	resets_inactivity: bool = True

	available_before_hello: bool = False

	#: True for a command that moves the user's machine, which an observe-only session refuses; the default
	#: is False, so a new mutating command must opt in.
	mutates_reader: bool = False

	#: GetLogHandler sets this False, so a getLog call is not its own anchor.
	marks_log: bool = True

	@abstractmethod
	def execute(self, ctx: SessionContext, request: protocol.Request) -> Any:
		"""Run the command and return its wire result, or raise to fail it."""

File: bridges/voiceover/Sources/VoiceOverBridgeDomain/Controllers/Commands/CommandHandler.swift
// ROLE: the interface every command handler implements, one handler per wire command.
// USED BY: the Session, which dispatches, and Registry, which builds the map.
// OWNS: `CommandError`, the failure a handler throws to answer with an error frame.
// A thrown error becomes an error frame; before `hello` it also ends the handshake.

import ScreenReaderWire

/// The longest any blocking command may hold the session thread; it must stay below
/// `SessionConfig.inactivityTimeout`, which is not refreshed when a handler returns.
public let maxPollTimeout: Double = 110.0

public struct CommandError: Error, Equatable, CustomStringConvertible {
	public let description: String

	public init(_ description: String) {
		self.description = description
	}
}

public protocol CommandHandler: AnyObject {
	/// `ping` alone returns false: it proves the process is alive, not that the agent is active.
	var resetsInactivity: Bool { get }

	var availableBeforeHello: Bool { get }

	/// Whether this command moves the user's machine; a mutating command must opt in, because the
	/// default of `false` lets an observe-only session run it.
	var mutatesReader: Bool { get }

	func execute(_ context: SessionContext, _ request: Request) throws -> any Encodable
}

public extension CommandHandler {
	var resetsInactivity: Bool { true }
	var availableBeforeHello: Bool { false }
	var mutatesReader: Bool { false }
}
